A48-xds-least-request-lb-policy.md

```diff
@@ -0,0 +1,225 @@
+inherit the logic instead. This would be similar to the "base balancer" already implemented in grpc-go
+[here](https://github.com/grpc/grpc-go/blob/03268c8ed29e801944a2265a82f240f7c0e1b1c3/balancer/base/balancer.go).
+The subchannel state handling used for the `round_robin` policy can be found in each language-specific implementation:
+
+* [Java](https://github.com/grpc/grpc-java/blob/c1e19af86dea0b9a8725969a95e116029397ad4d/core/src/main/java/io/grpc/util/RoundRobinLoadBalancer.java)
+* [C++](https://github.com/grpc/grpc/blob/79d684529d9db60aa2c5e82deb43e297a6818cfb/src/core/ext/filters/client_channel/lb_policy/round_robin/round_robin.cc)
+* [Go](https://github.com/grpc/grpc-go/blob/03268c8ed29e801944a2265a82f240f7c0e1b1c3/balancer/base/balancer.go)
+
+##### Aggregated Connectivity State
+
+The `least_request_experimental` policy will use the same heuristic for determining the aggregated
+connectivity state as the `round_robin` policy.
+The general rules in order of precedence are:
+
+1. If there is at least one subchannel with the state `READY`, the aggregated connectivity state is `READY`.
+2. If there is at least one subchannel with the state `CONNECTING` or `IDLE`,
+   the aggregated connectivity state is `CONNECTING`.
+3. If all subchannels have the state `TRANSIENT_FAILURE`, the aggregated connectivity state is `TRANSIENT_FAILURE`.
+
+##### Picker Behavior
+
+The picker should be given a list of all subchannels with the `READY` state.
+With this list, the picker should pick a subchannel based on the following pseudo-code:
+
+```
+candidate = null;
+for (int i = 0; i < choiceCount; ++i) {
+    sampled = subchannels[random_integer(0, subchannels.length)];
+    if (candidate == null) {
+        candidate = sampled;
+        continue;
+    }
+
+    if (sampled.active_requests < candidate.active_requests) {
+        candidate = sampled;
+    }
+}
+return candidate;
+```
+
+This pseudo-code is mirroring the
+[Envoy implementation](https://github.com/envoyproxy/envoy/blob/2443032526cf6e50d63d35770df9473dd0460fc0/source/common/upstream/load_balancer_impl.cc#L844-L865)
+
+##### Outstanding request counters
+
+The `least_request_experimental` policy will associate one outstanding request counter for each subchannel.
+These counters should additionally share the lifecycle with its corresponding subchannel.
+The counter for a subchannel should be atomically incremented by one after it has been successfully
+picked by the picker. Due to language-specifics, the increment may occur either before or during stream creation.
+In the `PickResult` for each language-specific implementation (
+[Java](https://github.com/grpc/grpc-java/blob/1f90e0e28d5628195cb1f861b73e45ed003f2973/api/src/main/java/io/grpc/LoadBalancer.java#L561-L571),
+[C++](https://github.com/grpc/grpc/blob/4567af504ed53cc8398e53bf1efd23f753d43bb8/src/core/ext/filters/client_channel/lb_policy.h#L178-L201),
```

Comment:
Note that I'm soon going to merge grpc/grpc#27947, which will change the API in C-core.  The new API handles an edge case that would have caused a problem for this design, in which the current API was not getting notified when a pick result was abandoned.

Reply:
Nice!
Then I would say it makes sense to wait until `subchannel_call_tracker` is available before implementing this.
Would be quite bad for the lb policy if outstanding requests are leaked.
We concluded in an earlier thread here that the Java `ClientStreamTracer` should be safe to use in this design.
Possibly we need to make sure the grpc-go callback also is safe for this design as well.

I'll can update the proposal here and reference that PR as a dependency.

Reply:
I've just merged that PR, so you can just refer to it as the current API now. :)

@dfawley tells me that the go callback is invoked even in the case where the pick is abandoned, so you should be able to do what you need in every language.